hazelcast/src/main/java/com/hazelcast/client/impl/connection/tcp/TcpClientConnectionManager.java

```diff
@@ -571,18 +571,24 @@ TcpClientConnection getOrConnect(@Nonnull Address address) {
             return connection;
         }
 
-        synchronized (resolveAddress(address)) {
-            // this critical section is used for making a single connection
-            // attempt to the given address at a time.
-            connection = getConnection(address);
-            if (connection != null) {
+        InetSocketAddress inetSocketAddress = resolveAddress(address);
+        try {
+            Object mutex = ConcurrencyUtil.getOrPutIfAbsent(mutexes, inetSocketAddress, key -> new Object());
```

Comment:
I'm thinking out loud.

What you could do is to create a CHM with address as key and a Future as value.

The future will either block if no value is set, or will return the correct connection.

You could add this future using e.g. a putIfAbcent. so that only 1 thread will wait for a connection and then set that value on the future.

Reply:
This way you prevent concurrent connection creations to the same address.

And you can use a CompletableFuture. Which also doesn't suffer from the monitor inflation problem.

And in theory a caller could timeut future.get with a timeout if needs to be.